### ISSUE TITLE
Referenced attributes set as required in generated code

### ISSUE THREAD

**Issue by NeenuAVarghese** [goadesign/goa#3112]
Referenced attributes set as required in generated code. Here is an example:
```
package design

import (
	. "goa.design/goa/v3/dsl"
)

var _ = API("calc", func() {
	Title("Calculator Service")
	Description("Service for multiplying numbers, a Goa teaser")
})

var _ = Service("calc", func() {
	Description("The calc service performs operations on numbers.")

	Method("multiply", func() {
		Payload(func() {
			Reference(CustomType)
			Attribute("a")
			Attribute("b")
			Attribute("c", Int)
			Attribute("d", Int)
		})

		Result(Int)

		HTTP(func() {
			GET("/multiply/{c}/{d}")
		})
	})

	Method("add", func() {
		Payload(func() {
			Attribute("two_vars", CustomType, func() {
				Required("a", "b")
			})
			Attribute("a")
			Attribute("b")
			Attribute("c", Int)
			Attribute("d", Int)
		})

		Result(Int)

		HTTP(func() {
			GET("/add/{c}/{d}")
		})
	})
})

var CustomType = Type("CustomType", func() {
	Attribute("a", CustomType2, "Left operand")
	Attribute("b", CustomType2, "Right operand")
})

var CustomType2 = Type("CustomType2", Int, func() {
	Description("this is a custom type2 of Int")
})

```

Generated service code for above:
```
// Code generated by goa v3.7.14, DO NOT EDIT.
//
// calc service
//
// Command:
// $ goa gen github.com/goa-ex/design --output gen/

package calc

import (
	"context"
)

// The calc service performs operations on numbers.
type Service interface {
	// Multiply implements multiply.
	Multiply(context.Context, *MultiplyPayload) (res int, err error)
	// Add implements add.
	Add(context.Context, *AddPayload) (res int, err error)
}

// ServiceName is the name of the service as defined in the design. This is the
// same value that is set in the endpoint request contexts under the ServiceKey
// key.
const ServiceName = "calc"

// MethodNames lists the service method names as defined in the design. These
// are the same values that are set in the endpoint request contexts under the
// MethodKey key.
var MethodNames = [2]string{"multiply", "add"}

// AddPayload is the payload type of the calc service add method.
type AddPayload struct {
	TwoVars *CustomType
	A       *string
	B       *string
	C       *int
	D       *int
}

type CustomType struct {
	// Left operand
	A CustomType2
	// Right operand
	B CustomType2
}

// this is a custom type2 of Int
type CustomType2 int

// MultiplyPayload is the payload type of the calc service multiply method.
type MultiplyPayload struct {
	// Left operand
	A CustomType2
	// Right operand
	B CustomType2
	C *int
	D *int
}

```

Note: This happens only when you introduce the `add` method which uses attributes from `CustomType` *and* sets them as required.

**Comment by raphael** [comment]
Again because the attributes are in the path :)

**Comment by NeenuAVarghese** [comment]
Sorry, seems like I come up with the right examples. Here is an example where the attributes are used as `Params` in a HTTP Get request and they are marked as required in the generated code:
```
package design

import (
	. "goa.design/goa/v3/dsl"
)

var _ = API("calc", func() {
	Title("Calculator Service")
	Description("Service for multiplying numbers, a Goa teaser")
})

// JWTAuth defines a security scheme that uses JWT tokens.
var JWTAuth = JWTSecurity("jwt", func() {
	Scope("my.scope", "Enable building shutdown")
})

var _ = Service("calc", func() {
	Description("The calc service performs operations on numbers.")
	Security(JWTAuth, func() {
		Scope("my.scope")
	})
	Method("multiply", func() {
		Payload(func() {
			Token("token", String)
			Reference(CustomType)
			Attribute("a")
			Attribute("b")
			Attribute("c", Int)
			Attribute("d", Int)
		})

		Result(Int)

		HTTP(func() {
			GET("/multiply/{c}/{d}")
			Param("a")
			Param("b")
		})
	})

	Method("add", func() {
		Payload(func() {
			Token("token", String)
			Attribute("two_vars", CustomType, func() {
				Required("a", "b")
			})
			Attribute("c", Int)
			Attribute("d", Int)
		})

		Result(Int)

		HTTP(func() {
			GET("/add/{c}/{d}")
		})
	})
})

var CustomType = Type("CustomType", func() {
	Attribute("a", CustomType2, "Left operand")
	Attribute("b", CustomType2, "Right operand")
})

var CustomType2 = Type("CustomType2", Int, func() {
	Description("this is a custom type2 of Int")
})

```

Generated code:

```
// Code generated by goa v3.7.14, DO NOT EDIT.
//
// calc service
//
// Command:
// $ goa gen github.com/goa-ex/design --output gen/

package calc

import (
	"context"

	"goa.design/goa/v3/security"
)

// The calc service performs operations on numbers.
type Service interface {
	// Multiply implements multiply.
	Multiply(context.Context, *MultiplyPayload) (res int, err error)
	// Add implements add.
	Add(context.Context, *AddPayload) (res int, err error)
}

// Auther defines the authorization functions to be implemented by the service.
type Auther interface {
	// JWTAuth implements the authorization logic for the JWT security scheme.
	JWTAuth(ctx context.Context, token string, schema *security.JWTScheme) (context.Context, error)
}

// ServiceName is the name of the service as defined in the design. This is the
// same value that is set in the endpoint request contexts under the ServiceKey
// key.
const ServiceName = "calc"

// MethodNames lists the service method names as defined in the design. These
// are the same values that are set in the endpoint request contexts under the
// MethodKey key.
var MethodNames = [2]string{"multiply", "add"}

// AddPayload is the payload type of the calc service add method.
type AddPayload struct {
	Token   *string
	TwoVars *CustomType
	C       *int
	D       *int
}

type CustomType struct {
	// Left operand
	A CustomType2
	// Right operand
	B CustomType2
}

// this is a custom type2 of Int
type CustomType2 int

// MultiplyPayload is the payload type of the calc service multiply method.
type MultiplyPayload struct {
	Token *string
	// Left operand
	A CustomType2
	// Right operand
	B CustomType2
	C *int
	D *int
}

```

**Comment by raphael** [comment]
I looked into this and the problem isn't the use of `Reference` but the fact that an attribute using `CustomType` adds required attributes.

The following would also cause attributes `A` and `B` to be required in the response for example:
```go
	Method("add", func() {
		Payload(func() {
			Attribute("two_vars", CustomType, func() {
				Required("a", "b")
			})
		})

		Result(CustomType)

		HTTP(func() {
			GET("/add/{c}/{d}")
		})
	})
```

Goa cannot simply duplicate the type when defining the `two_vars` attribute because that would break any recursive definition (duping the attribute would "break" the reference used to create the recursive type). Consider:

```go
var _ = Type("Child", func() {
    Attribute("Parent", "Parent", func() {
       Required("Name")
   })
})

var _ = Type("Parent", func() {
    Attribute("Child", "Child")
    Attribute("Name", String)
})
```

Duping the type `Parent` when defining the `Parent` attribute of the `Child` type would result in the dupe having no attribute (because attributes like `Name` are added to the original type reference).

TL;DR this is working as designed even if counter-intuitive: the type reference needs to point to the original type and so adding a required validation modifies the original type. In this particular case there could be another type that leverages `Extend` and adds the validations:

```
var CustomTypeWithRequired = Type("CustomTypeWithRequired", func() {
    Extend("CustomType")
    Required("a", "b")
})
```

Definitions that need the attributes to be required can then use `CustomTypeWithRequired` instead.
